File: solutions/javascript/src/algorithm/binarytree/Exercise3.js
import DynamicBinaryTreeADT from "../../structure/dynamic/DynamicBinaryTreeADT.js";

function isPrefix(a, b) {
  if (!a || a.isEmpty()) {
    return true;
  }
  if (!b || b.isEmpty()) {
    return false;
  }
  if (a.getRoot() !== b.getRoot()) {
    return false;
  }
  return isPrefix(a.getLeft(), b.getLeft()) && isPrefix(a.getRight(), b.getRight());
}


// Prueba
const tree1 = new DynamicBinaryTreeADT();
tree1.add(10);
tree1.add(5);

const tree2 = new DynamicBinaryTreeADT();
tree2.add(10);
tree2.add(5);
tree2.add(15);

console.log("Tree1 es prefijo de Tree2? ", isPrefix(tree1, tree2)); // true
console.log("Tree2 es prefijo de Tree1? ", isPrefix(tree2, tree1)); // false

File: solutions/python/algorithm/binarytree/Exercise3.py
from structure.dynamic.DynamicBinaryTreeADT import DynamicBinaryTreeADT


def is_prefix(a, b):
    if a is None or a.isEmpty():
        return True
    if b is None or b.isEmpty():
        return False
    if a.getRoot() != b.getRoot():
        return False
    return is_prefix(a.getLeft(), b.getLeft()) and is_prefix(a.getRight(), b.getRight())


# Prueba
tree1 = DynamicBinaryTreeADT()
tree1.add(10)
tree1.add(5)

tree2 = DynamicBinaryTreeADT()
tree2.add(10)
tree2.add(5)
tree2.add(15)

print("Tree1 es prefijo de Tree2? ", is_prefix(tree1, tree2))  # True
print("Tree2 es prefijo de Tree1? ", is_prefix(tree2, tree1))  # False